Removes Raft#join

Zelldon opened the pull request:
## Description

blocked by https://github.com/zeebe-io/zeebe/pull/6142
<!-- Please explain the changes you made here. -->

**Benchmark**
I would say it looks good. At the beginning we had a broker which was leader for all partition, which is the reason we see a bit degraded performance.

![general](https://user-images.githubusercontent.com/2758593/105451600-1c46c700-5c7d-11eb-97c0-db36d5642333.png)

Later an restart happened, because of oom (the new logSegment config of 128MB and beeing leader for all partitions). Restart seem to work as well. 
![general2](https://user-images.githubusercontent.com/2758593/105451598-1b159a00-5c7d-11eb-8eb1-8166c3aedf4e.png)

I see no issue in atomix sections.

![atomix](https://user-images.githubusercontent.com/2758593/105451601-1c46c700-5c7d-11eb-9638-75a1d27e9ddd.png)

![atomix2](https://user-images.githubusercontent.com/2758593/105451900-a1ca7700-5c7d-11eb-8f79-5daea8758fef.png)


## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #5292 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [x] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

Zelldon commented on atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftNamespaces.java:
```diff
@@ -46,18 +46,18 @@
-          .register(Void.class) // JoinRequest
-          .register(Void.class) // JoinResponse
-          .register(Void.class) // LeaveRequest
-          .register(Void.class) // LeaveResponse
```
As requested in the other review I removed the types here @deepthidevaki

Zelldon commented on atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/RaftClusterContext.java:
```diff
@@ -130,115 +112,56 @@ public DefaultRaftMember getMember(final MemberId id) {
     }
 
-    if (configuration == null) {
+    bootstrapFuture = new CompletableFuture<>();
```
True I missed that sorry

Zelldon replied:
@deepthidevaki please see my latest commit. I added a consistency check and test to verify that this works.

deepthidevaki reviewed (commented):
Thanks @Zelldon  I have some minor comments. Please take a look and see if you would like to apply them.

Zelldon commented on atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/RaftClusterContext.java:
```diff
@@ -50,30 +40,23 @@
   private volatile Configuration configuration;
-  private volatile Scheduled joinTimeout;
-  private volatile CompletableFuture<Void> joinFuture;
+  private volatile CompletableFuture<Void> bootstrapFuture;
```
I can do. I actually wanted to set config to final, but it is overwritten in same cases. Can we talk about that? I'm wondering whether these are necessary. For example here configure is called https://github.com/zeebe-io/zeebe/blob/develop/atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/DefaultRaftMember.java#L186 This is done on promote or denote a member to a certain state, like from PASSIVE to ACTIVE. 

Do we want to support the promote/demote stuff ? (\cc @npepinpe )

Zelldon replied:
I would propose that we remove these configuration changes as well.

Zelldon replied:
Ok let's do the following. I will use here atomic references instead of volatile and create a follow up issue to remove promote and demote, so we can make further progress and merge this PR already

Zelldon replied:
https://github.com/zeebe-io/zeebe/issues/6212

Zelldon commented:
@deepthidevaki I refactored a *bit* the RaftClusterContext to fix the volatile warnings and remove some other Sonar hints. Furthermore I updated the error messages, let me know what you think

deepthidevaki reviewed (approved):
Thanks. LGTM :+1:

Zelldon commented:
bors r+
